core/include/drivers/gpio.h

```diff
@@ -40,4 +66,140 @@ struct gpio_ops {
 			      enum gpio_interrupt ena_dis);
 };
 
+struct gpio {
+	struct gpio_chip *chip;
+	uint32_t dt_flags;
+	unsigned int pin;
+};
+
+static inline bool gpio_ops_is_valid(const struct gpio_ops *ops)
+{
+	return ops->set_direction && ops->get_direction && ops->get_value &&
+	       ops->set_value;
+}
+
+static inline void gpio_set_direction(struct gpio *gpio, enum gpio_dir dir)
+{
+	gpio->chip->ops->set_direction(gpio->chip, gpio->pin, dir);
+}
+
+static inline enum gpio_dir gpio_get_direction(struct gpio *gpio)
+{
+	return gpio->chip->ops->get_direction(gpio->chip, gpio->pin);
+}
+
+static inline void gpio_set_value(struct gpio *gpio, enum gpio_level value)
+{
+	if (gpio->dt_flags & GPIO_ACTIVE_LOW)
+		value = !value;
+
+	gpio->chip->ops->set_value(gpio->chip, gpio->pin, value);
+}
+
+static inline enum gpio_level gpio_get_value(struct gpio *gpio)
+{
+	enum gpio_level value = GPIO_LEVEL_LOW;
+
+	value = gpio->chip->ops->get_value(gpio->chip, gpio->pin);
+
+	if (gpio->dt_flags & GPIO_ACTIVE_LOW)
+		value = !value;
+
+	return value;
+}
+
+#if defined(CFG_DT) && defined(CFG_DRIVERS_GPIO)
+
+/**
+ * gpio_get_dt_two_cells - Typedef of function to get gpio controller from
+ * devicetree properties
+ *
+ * @a: Pointer to devicetree description of the gpio controller to parse
+ * @res: Output result code of the operation:
+ *	TEE_SUCCESS in case of success
+ *	TEE_ERROR_DEFER_DRIVER_INIT if gpio controller is not initialized
+ *	Any TEE_Result compliant code in case of error.
+ *
+ * Returns a struct gpio pointer pointing to a gpio controller matching
```

Comment:
s/a gpio controller/a GPIO instance/

We don't get a GPIO controller but rather a GPIO (`struct gpio`) from its GPIO controller (`struct gpio_chip`), no? Feel free to contradict me, I may have a bad/wrong habbit/use for these wordings.